Unable to log into gizmodo.com with tracking prevention enabled

szewai opened the pull request:
#### eb1f7a4e8a5e59ed10780be8869e34b5b6289b90
<pre>
Unable to log into gizmodo.com with tracking prevention enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=266442">https://bugs.webkit.org/show_bug.cgi?id=266442</a>
<a href="https://rdar.apple.com/106782128">rdar://106782128</a>

Reviewed by John Wilander.

Update the quirk to make authentication flow of gizmodo.com work by:
1. Invoking requestStorageAccess on behalf of kinja.com when user starts authentication flow by clicking user profile
button (console log is added to make user and developer aware of the quirk).
2. Adjusting heuristics of detecting user profile button by finding target class on its ancestors, instead of only on
the element itself.
3. Skip checking user interaction on kinja.com in requestStorageAccess.

* Source/WebCore/page/Quirks.cpp:
(WebCore::elementHasClassInClosestAncestors):
(WebCore::isStorageAccessQuirkDomainAndElement):
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::storageAccessQuirks):

Canonical link: <a href="https://commits.webkit.org/272146@main">https://commits.webkit.org/272146@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df36a63520af6dd8eacfd5c269e206725286e265

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32217 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33061 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27635 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31268 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11570 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6498 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27551 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30872 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7777 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27367 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6817 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27222 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34397 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27848 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32966 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4922 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30792 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8544 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7269 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7539 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7363 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1091,7 +1091,26 @@ bool Quirks::isMicrosoftTeamsRedirectURL(const URL& url)
     return url.host() == "teams.microsoft.com"_s && url.query().toString().contains("Retried+3+times+without+success"_s);
 }
 
-// microsoft.com rdar://72453487
```
Could you move the comment please? back underneath :)

szewai replied:
The radar is for "Can't login to Skype from Microsoft Outlook account in Safari" (https://trac.webkit.org/changeset/271074/webkit), which is fixed by:
```
// Skype case.
--
// FIXME(220105): Remove this quirk once Skype under outlook.live.com completes their login flow redesign.
if (url.host() == "outlook.live.com"_s) {
return element.hasClass()
&& (element.classNames().contains("_3ioEp2RGR5vb0gqRDsaFPa")
\|\| element.classNames().contains("_2Am2jvTaBz17UJ8XnfxFOy"));
}
```
and https://bugs.webkit.org/show_bug.cgi?id=220105 is already fixed, so I just dropped it.

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1191,7 +1215,6 @@ Quirks::StorageAccessResult Quirks::triggerOptionalStorageAccessQuirk(Element& e
         HashSet<RegistrableDomain> set;
         set.add(RegistrableDomain::uncheckedCreateFromRegistrableDomainString("avclub.com"_s));
-        set.add(RegistrableDomain::uncheckedCreateFromRegistrableDomainString("gizmodo.com"_s));
         set.add(RegistrableDomain::uncheckedCreateFromRegistrableDomainString("deadspin.com"_s));
         set.add(RegistrableDomain::uncheckedCreateFromRegistrableDomainString("jalopnik.com"_s));
         set.add(RegistrableDomain::uncheckedCreateFromRegistrableDomainString("jezebel.com"_s));
```
All these sites are part of the kinja network? 
They may exhibit the same issue?
Was it checked?

karlcow replied:
Though Jezebel has been abandoned and taken by another group.

Let's see. Same network than Gizmodo.

* https://www.avclub.com
* https://deadspin.com
* https://gizmodo.com
* https://jalopnik.com
* https://kotaku.com
* https://qz.com
* https://www.theroot.com
* https://thetakeout.com
* https://www.theonion.com
* https://theinventory.com

Interesting clicking the Jezebel login button still open the kinja network of sites.
https://jezebel.com
![Screenshot 2023-12-15 at 16 04 01](https://github.com/WebKit/WebKit/assets/505230/28199cbc-4ef1-4ca5-a92e-24b592a757f6)

szewai replied:
This radar is about gizmodo.com, so I limit the change to gizmodo.
As you said, other sites need to be evaluated case by case, i.e. whether the old quirk still works, whether they belong to the same organization nowadays, whether they have different UI elements, etc, so I don't think they belong to the same change. We could do in a followup.

johnwilander replied:
We have expressed in multiple standards conversations that the limit of domains under which a same-party SSO domain gets cookie access should be 5.

karlcow reviewed (changes requested):
@szewai Thanks for working on this.
A couple of comments related to the kinja network.
and one related to a comment which has been scraped.

johnwilander commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1111,6 +1130,10 @@ static bool isStorageAccessQuirkDomainAndElement(const URL& url, const Element&
     }
 
+    // Gizmodo login case: rdar://106782128.
+    if (url.host() == "gizmodo.com"_s)
```
Are we sure the host will always be gizmodo.com? Or should we look at the registrable domain to also cover subdomains?

szewai replied:
> We have expressed in multiple standards conversations that the limit of domains under which a same-party SSO domain gets cookie access should be 5.

Do you suggest a change here? This is not part of the patch and it's added in https://trac.webkit.org/changeset/263383/webkit for the old quirk, so I don't know much about it.

> Are we sure the host will always be gizmodo.com? Or should we look at the registrable domain to also cover subdomains?

I tested only on gizmodo.com for this radar, and it's "gizmodo.com".

johnwilander replied:
> > We have expressed in multiple standards conversations that the limit of domains under which a same-party SSO domain gets cookie access should be 5.
> 
> Do you suggest a change here? This is not part of the patch and it's added in https://trac.webkit.org/changeset/263383/webkit for the old quirk, so I don't know much about it.

No change requested. I just mentioned this since Karl brought up a large set of domains.

> > Are we sure the host will always be gizmodo.com? Or should we look at the registrable domain to also cover subdomains?
> 
> I tested only on gizmodo.com for this radar, and it's "gizmodo.com".

OK. We can change the quirk if we learn of any subdomain use.

johnwilander reviewed (commented):
Does this quirk invoke the regular Storage Access prompt to let the user decide?

johnwilander reviewed (approved):
r=me

szewai committed:
Unable to log into gizmodo.com with tracking prevention enabled
https://bugs.webkit.org/show_bug.cgi?id=266442
rdar://106782128

Reviewed by John Wilander.

Update the quirk to make authentication flow of gizmodo.com work by:
1. Invoking requestStorageAccess on behalf of kinja.com when user starts authentication flow by clicking user profile
button (console log is added to make user and developer aware of the quirk).
2. Adjusting heuristics of detecting user profile button by finding target class on its ancestors, instead of only on
the element itself.
3. Skip checking user interaction on kinja.com in requestStorageAccess.

* Source/WebCore/page/Quirks.cpp:
(WebCore::elementHasClassInClosestAncestors):
(WebCore::isStorageAccessQuirkDomainAndElement):
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::storageAccessQuirks):

Canonical link: https://commits.webkit.org/272146@main
